content/scripts/singletons/audio_player.gd

```diff
@@ -0,0 +1,133 @@
+		_intro_file_path = _random_file(Array(_song_data['song_intros']))
+		$music_intro.stream = load(_intro_file_path)
+	if play_ending:
+		_current_song_queue.append("ending")
+		_ending_file_path = _random_file(Array(_song_data['song_endings']))
+		$music_ending.stream = load(_ending_file_path)
+	_middle_file_path = _get_song_data(song_name)['song_middle']
+	$music_middle.stream = load(_random_file(Array(_middle_file_path)))
+	
+	#start playing by queue, for now we only chceck if music was marked to play with intro or not and play intro if it is needed.
+	if _current_song_queue.has("intro"):
+		$music_intro.play()
+	else:
+		$music_middle.play()
+# getting all information about song files and sfx
+func _get_song_data(song_name:String) -> Dictionary:
+	var _song_data = {}
+	var _s = load("res://content/audio/%s/init.gd" % song_name).new()
+	_song_data['song_intros'] = _s.song_intros
+	_song_data['song_middle'] = _s.song_middle
+	_song_data['song_endings'] = _s.song_endings
+	_song_data['song_full'] = _s.song_full
+	return _song_data
+
+func _get_sfx_data(song_name:String) -> Dictionary:
+	var _sfx_data = {}
+	var _s = load("res://content/audio/%s/init.gd" % song_name).new()
+	_sfx_data['sfx_file'] = _s.sfx_file
+	return _sfx_data
+#randomize file for intros and endings
+func _random_file(file_array:Array) -> String:
+	randomize()
+	var _file = file_array[randi() % file_array.size()]
+	return _file
+
+func _on_music_intro_finished() -> void:
+	print("Intro finished")
+	_current_song_queue.remove(_current_song_queue.find("intro")) #removing "intro"
+	$music_intro.stop()
+	$music_middle.play() #everything was preloaded previously, so for now we only call play method
+
+
+func _on_music_middle_finished() -> void:
+	if _middle_loop:
+		$music_middle.play()
+	else:
+		_current_song_queue.remove(_current_song_queue.find("middle")) #removing "intro"
+		# for now we check if we have to play ending, if so, ending will be played
+		if _current_song_queue.has("ending"):
+			$music_ending.play()
+		else:
+			_current_song_queue = [] #clear the song queue
+
+
+func _on_music_ending_finished() -> void:
+	_current_song_queue = [] #clear the song queue
+	$music_ending.stop() # just make sure ending won't loop
+
+#if we want to play full song from full file
+func play_full_song(song_idx:int) -> void:
+	var song_name = _song_indexes[str(song_idx)]
+	var _song_data = _get_song_data(song_name)
+	var _full_song_file = _random_file(Array(_song_data['song_full']))
+	$music_full.stream = load(_full_song_file)
+	var _ending_file_path = _random_file(Array(_song_data['song_endings']))
+	$music_ending.stream = load(_ending_file_path)
+	$music_full.play()
+
+
+func _on_music_full_finished() -> void:
+	pass # Replace with function body.
+
+func stop_song() -> void:
+	$music_middle.stop()
+	$music_intro.stop()
+	$music_ending.stop()
+	$music_full.stop()
+
+func pause_song(paused:bool=true) -> void:
+	$music_middle.stream_paused = paused
+	$music_intro.stream_paused = paused
```

Comment:
imo too many variables with dynamic node resolution
instead of calling get_node(xxx) ($xxx is just a syntax sugar), cache it at the start of the singleton
eg. 
onready var music_intro = $music_intro

it's overall faster + you don't run into issues with name changes

Reply:
done, thanks